libkbfs/journal_server.go

```diff
@@ -359,11 +430,49 @@ func (j *JournalServer) JournalStatus(tlfID TlfID) (TLFJournalStatus, error) {
 	return tlfJournal.getJournalStatus()
 }
 
+// shutdownExistingJournals shuts down all write journals, sets the
+// current UID and verifying key to zero, and returns once all
+// shutdowns are complete. It is safe to call multiple times in a row,
+// and once this is called, EnableExistingJournals may be called
+// again.
+func (j *JournalServer) shutdownExistingJournals(ctx context.Context) {
+	j.log.CDebugf(ctx, "Shutting down existing journals")
+
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	for _, tlfJournal := range j.tlfJournals {
+		tlfJournal.shutdown()
+	}
+
+	// Shutting down atomically is more important than respecting
+	// context cancellation.
+	waitCtx := context.Background()
+	for tlfID, tlfJournal := range j.tlfJournals {
+		err := tlfJournal.wait(waitCtx)
```

Comment:
Ah, okay, I see the problem now.

I made `tlfJournal.shutdown` wait for the background thread to finish before returning, which seems easier to reason about, anyway. Then the wait isn't needed.